respy/conditional_draws.py

```diff
@@ -54,95 +55,128 @@ def create_draws_and_log_prob_wages(
-    extended_cholcovs_t[:, 1:, 1:] = shocks_cholesky.T
-    meas_sds = meas_error_sds[choices]
-
-    # Note: ``states`` and ``extended_cholcovs_t`` are changed in-place.
-    log_prob_wages = kalman_update(
-        states, measurements, extended_cholcovs_t, meas_sds, choices
+    updated_means, log_prob_wages = update_mean_and_evaluate_likelihood(
+        observed_shocks, cov, choices, meas_sds
     )
 
-    draws = np.matmul(base_draws[periods], extended_cholcovs_t[:, 1:, 1:])
-    draws += states.reshape(n_obs, 1, n_choices)
-    draws[:, :n_wages] = np.clip(np.exp(draws[:, :n_wages]), 0.0, HUGE_FLOAT)
+    if is_meas_error:
+        updated_chols = update_cholcov_with_measurement_error(shocks_cholesky, meas_sds)
+    else:
+        updated_chols = update_cholcov(shocks_cholesky, n_wages)
+
+    chol_indices = np.where(np.isfinite(log_wages_observed), choices, n_wages)
+    draws = calculate_conditional_draws(
+        base_draws[periods], updated_means, updated_chols, chol_indices
+    )
 
     return draws, log_prob_wages
 
 
 @guvectorize(
-    ["f8[:], f8, f8[:, :], f8, u2, f8[:]"],
-    "(n_states), (), (n_choices_plus_one, n_choices_plus_one), (), () -> ()",
+    ["f8, f8[:, :], u2, f8[:], f8[:], f8[:]"],
+    "(), (n_choices, n_choices), (), (n_wages) -> (n_choices), ()",
+    nopython=True,
 )
-def kalman_update(
-    state, measurement, extended_cholcov_t, meas_sd, choice, log_prob_wage
+def update_mean_and_evaluate_likelihood(
+    shock, cov, choice, meas_sds, updated_mean, loglike
 ):
-    """Make a Kalman update and evaluate log likelihood of wages.
+    """Update mean and evaluate likelihood.
+
+    Calculate the conditional mean of shocks after observing one shock
+    and evaluate the likelihood of the observed shock.
+
+    Parameters
+    ----------
+    shock : float
+        The observed shock. NaN for individuals where no shock was observed.
+    cov : np.ndarray
+        Unconditional covariance matrix of the shocks
+    choice : int
+        The observed choice
+    meas_sds : np.ndarray
+        1d array of length n_choices with standard errors of measurement errors.
+
+    Returns
+    -------
+    updated_mean : np.ndarray
+        Conditional mean of shocks, given the observed shock. Contains the observed
+        shock in the corresponding position even in the degenerate case of no
+        measurement error. Has length n_choices.
+    loglike : float
+        log likelihood of observing the observed shock. 0 if no shock was observed.
 
-    This function is a specialized Kalman filter in the following ways:
+    """
+    dim = len(cov)
+    invariant = np.log(1 / (2 * np.pi) ** 0.5)
 
-    - All measurements just measure one latent factor
-    - All factor loadings are equal to 1
-    - All initial states are assumed to be equal to 0
+    if np.isfinite(shock):
+        sigma_squared = cov[choice, choice] + meas_sds[choice] ** 2
+        sigma = np.sqrt(sigma_squared)
```

Comment:
Probably, this is already heavily optimized, but some comments.
- Why do you put ``invariant`` outside, but ``sigma_squared`` inside? Does it make sense to defer the one calculation, but do the other?
- you do not actually need ``sigma`` but ``log(sqrt(sigma_squared))``. Does it make sense to make it more compact? Or does it make the function slower?

Reply:
It was just faster when I benchmarked it once.

Reply:
Then, it is fine.